### PR TITLE
Introducing signals in nailgun 0.27.0

### PR DESCRIPTION
**Pull request by rochacbruno**
- Add blinker as an optional requirement, if not provided signals will be ignored unless there is a handler trying to connect, in that case it will raise an exception informing that blinker needs to be installed.
- Add documentation for signals module
- Add test_signals to test only that all signals are properly emitted

```python
from nailgun import entities, signals

@signals.pre_create.connect
def pre_create_handler(sender, create_missing):
    if isinstance(sender, entities.Organization):
        print("Org will be created, lets assign its domain")
        sender.domain = "http://example.com"

@signals.post_create.connect
def post_create_handler(sender, entity):
    if isinstance(entity, entities.Organization):
        print("Org {e.id} was created with domain {e.domain}!!!".format(e=entity))


>>> org = Organization(cfg, id=1).create()
>>> 'Org will be created, lets assign its domain' 
>>> 'Org 1 was created with domain http://example.com!!!'
```

### CODE REVIEW

File: tests/test_signals.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+if version_info < (3, 4):
+    from unittest2 import TestCase
+else:
+    from unittest import TestCase
+
+ORG_DATA = {"id": 1, "domain": "http://example.com"}
+
+
+class FakeResponse(object):
+    '''Faking here cause we only need to test signals'''
```
Comment (elyezer) [inline tests/test_signals.py]: We use `"` for docstring, can you please follow the code style? Same applies for other docstrings.

File: tests/test_signals.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+from sys import version_info
+from mock import patch
+from nailgun import entities, config, signals, client
+
+if version_info < (3, 4):
+    from unittest2 import TestCase
+else:
+    from unittest import TestCase
+
+ORG_DATA = {"id": 1, "domain": "http://example.com"}
```
Comment (elyezer) [inline tests/test_signals.py]: We use single quote for string in general, please update the code.

File: tests/test_signals.py (hunk 1 above)
Comment (elyezer) [inline tests/test_signals.py]: Why not use mock for the response?

**Comment by elyezer** [comment]
It looks good to me. I really like the signals approach specially by making them optional.

File: requirements-dev.txt [hunk 3]
```diff
@@ -19,3 +19,6 @@ twine
 
 # For code coverage
 coveralls
+
+# For signals
+blinker
```
Comment (elyezer) [inline requirements-dev.txt]: :+1:

File: tests/test_signals.py (hunk 2 above)
Comment (rochacbruno) [inline tests/test_signals.py]: Done!

File: tests/test_signals.py (hunk 1 above)
Comment (rochacbruno) [inline tests/test_signals.py]: Done!
Do you have an example of using mock for response making status_code returning 200 and json returning some valid data?

**Comment by elyezer** [comment]
ACK, thank you @rochacbruno

File: nailgun/entity_mixins.py [hunk 4]
```diff
@@ -839,7 +847,10 @@ def create(self, create_missing=None):
             on the current object.
 
         """
-        return self.read(attrs=self.create_json(create_missing))
+        signals.pre_create.send(self, create_missing=create_missing)
+        entity = self.read(attrs=self.create_json(create_missing))
+        signals.post_create.send(self, entity=entity)
+        return entity
```
Comment (sthirugn) [inline nailgun/entity_mixins.py]: Does Read action really need a signal?

File: nailgun/entity_mixins.py [hunk 5]
```diff
@@ -1245,6 +1261,9 @@ def search(self, fields=None, query=None, filters=None):
         ]
         if filters is not None:
             entities = self.search_filter(entities, filters)
+
+        signals.post_search.send(self, entities=entities, fields=fields,
+                                 query=query, filters=filters)
```
Comment (sthirugn) [inline nailgun/entity_mixins.py]: Do you think of any scenario in which signals will be useful for a search operation?

**Comment by sthirugn** [comment]
ACK pending questions

File: nailgun/entity_mixins.py (hunk 4 above)
Comment (rochacbruno) [inline nailgun/entity_mixins.py]: This is not a **read** this is **create** take a look https://github.com/rochacbruno/nailgun/blob/add_signals/nailgun/entity_mixins.py#L824-L850 explanation on docstring (looks like **.read** actually creates with that attrs create_missing.

File: nailgun/entity_mixins.py (hunk 5 above)
Comment (rochacbruno) [inline nailgun/entity_mixins.py]: I imagine a scenario where one would like to apply a filter to every search in batch,

```python
@signals.pre_search.connect
def add_filters_to_every_search(sender, fields, query, filters):
    filters.append(#filters_here)
```

With that any **.search** performed will automatically get that filters added.

The post_search may be useful to record logs, statistics etc

There is no overhead on emitting signals, but it looks like uncommon scenarios, if agreed I can remove the search signals.

```

Comment (sthirugn) [inline nailgun/entity_mixins.py]: np, we can leave it if harmless

**Comment by sthirugn** [comment]
and pypi is now refreshed with version 0.27.0

**Pull request merged**